docs/_docs/02_features/keep-your-terragrunt-architecture-dry.md

```diff
@@ -0,0 +1,345 @@
+However, it is cumbersome to copy paste these settings across all three environments.
+
+To solve this, you can use [multiple include blocks]({{site.baseurl}}/docs/reference/config-blocks-and-attributes#include).
+
+### Using include to DRY common Terragrunt config
+
+Suppose your `qa/app/terragrunt.hcl` configuration looks like the following:
+
+```hcl
+include {
+  path = find_in_parent_folders()
+}
+
+terraform {
+  source = "github.com/<org>/modules.git//app?ref=v0.1.0"
+}
+
+dependency "vpc" {
+  config_path = "../vpc"
+}
+
+dependency "mysql" {
+  config_path = "../mysql"
+}
+
+inputs = {
+  env            = "qa"
+  basename       = "example-app"
+  vpc_id         = dependency.vpc.outputs.vpc_id
+  subnet_ids     = dependency.vpc.outputs.subnet_ids
+  mysql_endpoint = dependency.mysql.outputs.endpoint
+}
+```
+
+In this example, the only thing that is different between the environments is the `env` input variable. This means that
+except for one line, everything in the config is duplicated across `prod`, `qa`, and `stage`.
+
+To DRY this up, we will introduce a new folder called `_env` which will contain the common configurations across the
+three environments (we prefix with `_` to indicate that this folder doesn't contain deployable configurations):
+
+    └── live
+        ├── terragrunt.hcl
+        ├── _env
+        │   ├── app.hcl
+        │   ├── mysql.hcl
+        │   └── vpc.hcl
+        ├── prod
+        │   ├── app
+        │   │   └── terragrunt.hcl
+        │   ├── mysql
+        │   │   └── terragrunt.hcl
+        │   └── vpc
+        │       └── terragrunt.hcl
+        ├── qa
+        │   ├── app
+        │   │   └── terragrunt.hcl
+        │   ├── mysql
+        │   │   └── terragrunt.hcl
+        │   └── vpc
+        │       └── terragrunt.hcl
+        └── stage
+            ├── app
+            │   └── terragrunt.hcl
+            ├── mysql
+            │   └── terragrunt.hcl
+            └── vpc
+                └── terragrunt.hcl
+
+In our example, the contents of `_env/app.hcl` would look like the following:
+
+```hcl
+terraform {
+  source = "github.com/<org>/modules.git//app?ref=v0.1.0"
+}
+
+dependency "vpc" {
+  config_path = "../vpc"
+}
+
+dependency "mysql" {
+  config_path = "../mysql"
+}
+
+inputs = {
+  basename       = "example-app"
+  vpc_id         = dependency.vpc.outputs.vpc_id
+  subnet_ids     = dependency.vpc.outputs.subnet_ids
+  mysql_endpoint = dependency.mysql.outputs.endpoint
+}
+```
+
+Note that everything is defined except for the `env` input variable. We now modify `qa/app/terragrunt.hcl` to include
+this alongside the root configuration by using multiple `include` blocks, significantly reducing our per
+environment configuration:
+
+```hcl
+include "root" {
+  path = find_in_parent_folders()
+}
+
+include "env" {
+  path = "${get_terragrunt_dir()}/../../_env/app.hcl"
+}
+
+inputs = {
+  env = "qa"
+}
+```
+
+### Using read\_terragrunt\_config to DRY parent configurations
```

Comment:
I mentioned this in the other PR, but we _probably_ want to encourage putting things that you typically change on an env-by-env basis in the env itself so it works better with CI / CD; and putting things that you would change globally (albeit rarely) into an `_env` folder. So for example, env name, instance type, number of instances, `ref` (version) are all things you may want to change in one env, but not another, and you may want CI / CD to be able to detect it based on the file you changed, so it makes more sense to define those params directly in `stage/app/terragrunt.hcl` for stage and `prod/app/terragrunt.hcl` for prod. For everything else, you can put it in `_env`, under the assumption that it changes very rarely, and you can somehow manually trigger CI / CD as necessary for that change.

That does make me think of an alternative approach, which might be nonsense, but I figured it's worth thinking through: would it make sense to version the `terragrunt.hcl` files in `_env`? Ignoring implementation details, imagine for a second that all the `_env` stuff is defined in `infrastructure-modules`; inside the `terragrunt.hcl` files in `_env`, it uses relative paths within `infrastructure-modules`; and the whole repo is versioned as usual. Then, in `infrastructure-live`, you `include` not from a `path` but from a versioned `source` parameter. That allows you to use different versions of `_env` in each actual env, and promote changes from one env to another. In that case, you _could_ define environment-specific values in `_env` as well (e.g., instance type, instance count, etc), as you'd bump the version number in `stage` or `prod` to pick up the new configs, and CI / CD would pick that up.

So, in effect, `infra-live` would become VERY minimal: just a bunch of `terragrunt.hcl` marker files that use `include` to pull in the actual configs from versioned URLs.

Reply:
> That does make me think of an alternative approach, which might be nonsense, but I figured it's worth thinking through: would it make sense to version the terragrunt.hcl files in _env? Ignoring implementation details, imagine for a second that all the _env stuff is defined in infrastructure-modules; inside the terragrunt.hcl files in _env, it uses relative paths within infrastructure-modules; and the whole repo is versioned as usual. Then, in infrastructure-live, you include not from a path but from a versioned source parameter. That allows you to use different versions of _env in each actual env, and promote changes from one env to another. In that case, you could define environment-specific values in _env as well (e.g., instance type, instance count, etc), as you'd bump the version number in stage or prod to pick up the new configs, and CI / CD would pick that up.

Yes I had this exact thought! This actually has another advantage of better handling variable updates across module versions. E.g., in the current setup, there is a limitation where the shared inputs are in a static file shared across `dev`, `stage`, and `prod`. What if a version bump removes a variable, or there is a new required variable, or an old variable type changed? Ideally, you would make the change in the shared variables in `_env`, but then that would break `stage` and `prod`.

Reply:
> I mentioned this in the other PR, but we probably want to encourage putting things that you typically change on an env-by-env basis in the env itself so it works better with CI / CD; and putting things that you would change globally (albeit rarely) into an _env folder. So for example, env name, instance type, number of instances, ref (version) are all things you may want to change in one env, but not another, and you may want CI / CD to be able to detect it based on the file you changed, so it makes more sense to define those params directly in stage/app/terragrunt.hcl for stage and prod/app/terragrunt.hcl for prod. For everything else, you can put it in _env, under the assumption that it changes very rarely, and you can somehow manually trigger CI / CD as necessary for that change.

Good point. I updated the docs to use `expose` instead of `read_terragrunt_config` for the module version, and updated `read_terragrunt_config` docs to be something more generic: c034ca9c79d780ff8aff08fadb710f76ed777dba

Reply:
> > That does make me think of an alternative approach, which might be nonsense, but I figured it's worth thinking through: would it make sense to version the terragrunt.hcl files in _env? Ignoring implementation details, imagine for a second that all the _env stuff is defined in infrastructure-modules; inside the terragrunt.hcl files in _env, it uses relative paths within infrastructure-modules; and the whole repo is versioned as usual. Then, in infrastructure-live, you include not from a path but from a versioned source parameter. That allows you to use different versions of _env in each actual env, and promote changes from one env to another. In that case, you could define environment-specific values in _env as well (e.g., instance type, instance count, etc), as you'd bump the version number in stage or prod to pick up the new configs, and CI / CD would pick that up.
> 
> Yes I had this exact thought! This actually has another advantage of better handling variable updates across module versions. E.g., in the current setup, there is a limitation where the shared inputs are in a static file shared across `dev`, `stage`, and `prod`. What if a version bump removes a variable, or there is a new required variable, or an old variable type changed? Ideally, you would make the change in the shared variables in `_env`, but then that would break `stage` and `prod`.

We should def experiment with this as the way to make the Ref Arch super DRY, versioned, etc.